File: clients/kmp/shared/src/commonTest/kotlin/org/wyrdsekai/app/engine/PhoneNodeTierTest.kt
package org.wyrdsekai.app.engine

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.launch
import kotlinx.coroutines.test.advanceTimeBy
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.yield
import org.wyrdsekai.app.engine.tier.*
import org.wyrdsekai.app.inference.InferenceClient
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue
import kotlin.test.assertFalse
import kotlin.test.assertNotNull

@OptIn(ExperimentalCoroutinesApi::class)
class PhoneNodeTierTest {

    private fun makeProbe(
        memoryMb: Long = 2000,
        battery: Int = 80,
        charging: Boolean = false,
        thermal: ThermalState = ThermalState.NOMINAL,
        wifi: Boolean = true,
    ) = object : ResourceProbe {
        override fun snapshot() = ResourceSnapshot(
            availableMemoryMb = memoryMb,
            totalMemoryMb = 4000,
            batteryPercent = battery,
            isCharging = charging,
            thermalState = thermal,
            hasWifi = wifi,
        )
    }

    private fun makeNode(
        scope: kotlinx.coroutines.CoroutineScope,
        tierManager: TierManager? = null,
    ) = PhoneNode(
        journal = InMemoryEventJournal(),
        vitalityStore = null,
        inferenceClient = InferenceClient(),
        inferenceBaseUrl = "http://test",
        scope = scope,
        tierManager = tierManager,
    )

    // ── Room definitions ────────────────────────────────────────────────

    @Test
    fun roomsForTierAreCumulative() {
        val node = makeNode(
            scope = kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.Default),
        )

        // T0 and T1 both have study (companion's home base on phone)
        assertEquals(listOf("study"), node.roomsForTier(Tier.T0))
        assertEquals(listOf("study"), node.roomsForTier(Tier.T1))
        // T2 adds home + more rooms
        assertTrue(node.roomsForTier(Tier.T2).contains("study"))
        assertTrue(node.roomsForTier(Tier.T2).contains("home"))
        assertTrue(node.roomsForTier(Tier.T2).contains("terminal"))
        assertTrue(node.roomsForTier(Tier.T2).size > node.roomsForTier(Tier.T1).size)
        // T3 adds even more
        assertTrue(node.roomsForTier(Tier.T3).size > node.roomsForTier(Tier.T2).size)
    }

    @Test
    fun allRoomDefinitionsExist() {
        val node = makeNode(
            scope = kotlinx.coroutines.CoroutineScope(kotlinx.coroutines.Dispatchers.Default),
        )
        val allRoomIds = Tier.entries.flatMap { node.roomsForTier(it) }.toSet()

        for (roomId in allRoomIds) {
            assertNotNull(
                PhoneNode.ROOM_DEFINITIONS[roomId],
                "Room definition missing for '$roomId'"
            )
        }
    }

    @Test
    fun roomDefinitionsHaveExitsAndDescriptions() {
        for ((roomId, def) in PhoneNode.ROOM_DEFINITIONS) {
            assertTrue(def.name.isNotBlank(), "Room '$roomId' has empty name")
            assertTrue(def.description.isNotBlank(), "Room '$roomId' has empty description")
            assertTrue(def.zone.isNotBlank(), "Room '$roomId' has empty zone")
            // All rooms should have at least one exit (back to home, or home has exits to others)
            assertTrue(def.exits.isNotEmpty(), "Room '$roomId' has no exits")
        }
    }

    // ── Tier-aware boot (rooms only, no companion/inference) ────────────

    @Test
    fun bootAtT1CreatesStudyOnly() = runTest {
        val probe = makeProbe(memoryMb = 1500, wifi = false)
        val tierManager = TierManager(probe, scope = this)
        val node = makeNode(scope = this, tierManager = tierManager)

        node.startRoomsOnly()

        assertEquals(PhoneNode.State.RUNNING, node.state.value)
        assertEquals(Tier.T1, node.currentTier)
        assertTrue(node.activeRoomIds().contains("study"))
        assertFalse(node.activeRoomIds().contains("home"))
        assertFalse(node.activeRoomIds().contains("terminal"))

        node.stop()
    }

    @Test
    fun bootAtT2CreatesMultipleRooms() = runTest {
        val probe = makeProbe(memoryMb = 2500, wifi = true)
        val tierManager = TierManager(probe, scope = this)
        val node = makeNode(scope = this, tierManager = tierManager)

        node.startRoomsOnly()

        assertEquals(Tier.T2, node.currentTier)
        assertTrue(node.activeRoomIds().contains("study"))
        assertTrue(node.activeRoomIds().contains("home"))
        assertTrue(node.activeRoomIds().contains("terminal"))
        assertTrue(node.activeRoomIds().contains("dream-chamber"))
        assertTrue(node.activeRoomIds().contains("mailroom"))

        node.stop()
    }

    @Test
    fun bootWithoutTierManagerDefaultsToT1() = runTest {
        val node = makeNode(scope = this, tierManager = null)

        node.startRoomsOnly()

        assertEquals(PhoneNode.State.RUNNING, node.state.value)
        assertEquals(Tier.T1, node.currentTier)
        assertTrue(node.activeRoomIds().contains("study"))
        assertFalse(node.activeRoomIds().contains("home"))

        node.stop()
    }

    // ── Tier transitions ────────────────────────────────────────────────

    @Test
    fun promotionActivatesNewRooms() = runTest {
        val probe = makeProbe(memoryMb = 1500, wifi = false)
        val tierManager = TierManager(probe, scope = this)
        val node = makeNode(scope = this, tierManager = tierManager)

        node.startRoomsOnly()
        yield() // Let tier listener collector start
        assertEquals(Tier.T1, node.currentTier)
        assertEquals(1, node.activeRoomIds().size)

        // Promote to T2 — advanceTimeBy lets bootRoom's delay(100) complete for each room
        tierManager.forceTier(Tier.T2)
        advanceTimeBy(1000); yield()

        assertEquals(Tier.T2, node.currentTier)
        assertTrue(node.activeRoomIds().contains("terminal"))
        assertTrue(node.activeRoomIds().contains("dream-chamber"))

        node.stop()
    }

    @Test
    fun demotionPassivatesRooms() = runTest {
        val probe = makeProbe(memoryMb = 2500, wifi = true)
        val tierManager = TierManager(probe, scope = this)
        val node = makeNode(scope = this, tierManager = tierManager)

        node.startRoomsOnly()
        yield() // Let tier listener collector start
        assertEquals(Tier.T2, node.currentTier)
        assertTrue(node.activeRoomIds().size >= 5)

        // Demote to T1
        tierManager.forceTier(Tier.T1)
        advanceTimeBy(1000); yield()

        assertEquals(Tier.T1, node.currentTier)
        assertEquals(setOf("study"), node.activeRoomIds())
        assertTrue(node.passivatedRoomIds().contains("home"))
        assertTrue(node.passivatedRoomIds().contains("terminal"))
        assertTrue(node.passivatedRoomIds().contains("dream-chamber"))

        node.stop()
    }

    @Test
    fun reactivationAfterPromotion() = runTest {
        val probe = makeProbe(memoryMb = 2500, wifi = true)
        val tierManager = TierManager(probe, scope = this)
        val node = makeNode(scope = this, tierManager = tierManager)

        node.startRoomsOnly()
        yield() // Let tier listener collector start

        // Demote then re-promote
        tierManager.forceTier(Tier.T1)
        advanceTimeBy(1000); yield()
        assertTrue(node.passivatedRoomIds().contains("terminal"))

        tierManager.forceTier(Tier.T2)
        advanceTimeBy(1000); yield()
        assertTrue(node.activeRoomIds().contains("terminal"))
        assertFalse(node.passivatedRoomIds().contains("terminal"))

        node.stop()
    }

    @Test
    fun tierChangedEventEmitted() = runTest {
        val probe = makeProbe(memoryMb = 1500, wifi = false)
        val tierManager = TierManager(probe, scope = this)
        val node = makeNode(scope = this, tierManager = tierManager)

        val events = mutableListOf<PhoneNodeEvent>()
        val job = launch { node.notifications.collect { events.add(it) } }
        yield() // Let collector start

        node.startRoomsOnly()
        yield() // Let tier listener collector start

        tierManager.forceTier(Tier.T3)
        advanceTimeBy(1000); yield()

        assertTrue(events.any { it is PhoneNodeEvent.TierChanged })
        val tierEvent = events.filterIsInstance<PhoneNodeEvent.TierChanged>().first()
        assertEquals(Tier.T1, tierEvent.from)
        assertEquals(Tier.T3, tierEvent.to)

        job.cancel()
        node.stop()
    }
}

File: clients/rn/__tests__/engine/phone-node-tier.test.ts
import { PhoneNode, PhoneNodeEvent, ROOM_DEFINITIONS } from '../../src/engine/PhoneNode';
import type { ModelRole } from '../../src/inference/InferenceRouter';
import { InMemoryEventJournal } from '../../src/engine/persistence/InMemoryEventJournal';
import type { ResourceProbe, ResourceSnapshot, Tier } from '../../src/engine/tier/TierConfig';
import { TIER_ORDER } from '../../src/engine/tier/TierConfig';
import { TierManager } from '../../src/engine/tier/TierManager';
import type { ChatMessage, ChatResponse } from '../../src/inference/types';

const mockInference = {
  async complete(_role: ModelRole, _messages: ChatMessage[]): Promise<ChatResponse> {
    return { content: 'echo', promptTokens: 10, completionTokens: 5 };
  },
};

function makeProbe(overrides: Partial<ResourceSnapshot> = {}): ResourceProbe {
  return {
    snapshot: () => ({
      availableMemoryMb: 2000,
      totalMemoryMb: 4000,
      batteryPercent: 80,
      isCharging: false,
      thermalState: 'NOMINAL' as const,
      hasWifi: true,
      ...overrides,
    }),
  };
}

function makeNode(tierManager: TierManager | null = null): PhoneNode {
  return new PhoneNode(
    new InMemoryEventJournal(),
    null,
    mockInference,
    tierManager,
  );
}

describe('PhoneNode tier — room definitions', () => {
  it('roomsForTier are cumulative', () => {
    const node = makeNode();
    // T0/T1 now boot the Study (entry room) plus the Home room.
    expect(node.roomsForTier('T0')).toEqual(['study', 'home']);
    expect(node.roomsForTier('T1')).toEqual(['study', 'home']);
    expect(node.roomsForTier('T2')).toContain('home');
    expect(node.roomsForTier('T2')).toContain('terminal');
    expect(node.roomsForTier('T2').length).toBeGreaterThan(node.roomsForTier('T1').length);
    expect(node.roomsForTier('T3').length).toBeGreaterThan(node.roomsForTier('T2').length);
  });

  it('all room definitions exist', () => {
    const node = makeNode();
    const allRoomIds = new Set(TIER_ORDER.flatMap(t => node.roomsForTier(t)));
    for (const roomId of allRoomIds) {
      expect(ROOM_DEFINITIONS[roomId]).toBeDefined();
    }
  });

  it('room definitions have exits and descriptions', () => {
    for (const [roomId, def] of Object.entries(ROOM_DEFINITIONS)) {
      expect(def.name.length).toBeGreaterThan(0);
      expect(def.description.length).toBeGreaterThan(0);
      expect(def.zone.length).toBeGreaterThan(0);
      expect(def.exits.length).toBeGreaterThan(0);
    }
  });
});

describe('PhoneNode tier — boot', () => {
  it('boot at T1 creates study + home only', async () => {
    const probe = makeProbe({ availableMemoryMb: 1500, hasWifi: false });
    const tierManager = new TierManager(probe);
    const node = makeNode(tierManager);

    await node.startRoomsOnly();

    expect(node.state).toBe('running');
    expect(node.currentTier).toBe('T1');
    expect(node.activeRoomIds().has('home')).toBe(true);
    expect(node.activeRoomIds().has('terminal')).toBe(false);

    node.stop();
  });

  it('boot at T2 creates multiple rooms', async () => {
    const probe = makeProbe({ availableMemoryMb: 2500 });
    const tierManager = new TierManager(probe);
    const node = makeNode(tierManager);

    await node.startRoomsOnly();

    expect(node.currentTier).toBe('T2');
    expect(node.activeRoomIds().has('home')).toBe(true);
    expect(node.activeRoomIds().has('terminal')).toBe(true);
    expect(node.activeRoomIds().has('dream-chamber')).toBe(true);
    expect(node.activeRoomIds().has('mailroom')).toBe(true);

    node.stop();
  });

  it('boot without tier manager defaults to T1', async () => {
    const node = makeNode(null);

    await node.startRoomsOnly();

    expect(node.state).toBe('running');
    expect(node.currentTier).toBe('T1');
    expect(node.activeRoomIds().has('home')).toBe(true);

    node.stop();
  });
});

describe('PhoneNode tier — transitions', () => {
  it('promotion activates new rooms', async () => {
    const probe = makeProbe({ availableMemoryMb: 1500, hasWifi: false });
    const tierManager = new TierManager(probe);
    const node = makeNode(tierManager);

    await node.startRoomsOnly();
    expect(node.currentTier).toBe('T1');
    // T1 boots two rooms now: the Study and Home.
    expect(node.activeRoomIds().size).toBe(2);

    // Promote to T2
    tierManager.forceTier('T2');

    // Tier transition handler is synchronous (listener callback)
    // but bootRoom is async — wait for rooms to boot
    await delay(500);

    expect(node.currentTier).toBe('T2');
    expect(node.activeRoomIds().has('terminal')).toBe(true);
    expect(node.activeRoomIds().has('dream-chamber')).toBe(true);

    node.stop();
  });

  it('demotion passivates rooms', async () => {
    const probe = makeProbe({ availableMemoryMb: 2500 });
    const tierManager = new TierManager(probe);
    const node = makeNode(tierManager);

    await node.startRoomsOnly();
    expect(node.currentTier).toBe('T2');
    expect(node.activeRoomIds().size).toBeGreaterThanOrEqual(4);

    // Demote to T1
    tierManager.forceTier('T1');

    expect(node.currentTier).toBe('T1');
    expect(setEquals(node.activeRoomIds(), new Set(['study', 'home']))).toBe(true);
    expect(node.passivatedRoomIds().has('terminal')).toBe(true);
    expect(node.passivatedRoomIds().has('dream-chamber')).toBe(true);

    node.stop();
  });

  it('reactivation after promotion', async () => {
    const probe = makeProbe({ availableMemoryMb: 2500 });
    const tierManager = new TierManager(probe);
    const node = makeNode(tierManager);

    await node.startRoomsOnly();

    // Demote then re-promote
    tierManager.forceTier('T1');
    expect(node.passivatedRoomIds().has('terminal')).toBe(true);

    tierManager.forceTier('T2');
    await delay(500);

    expect(node.activeRoomIds().has('terminal')).toBe(true);
    expect(node.passivatedRoomIds().has('terminal')).toBe(false);

    node.stop();
  });

  it('tier changed event emitted', async () => {
    const probe = makeProbe({ availableMemoryMb: 1500, hasWifi: false });
    const tierManager = new TierManager(probe);
    const node = makeNode(tierManager);

    const events: PhoneNodeEvent[] = [];
    node.onEvent(e => events.push(e));

    await node.startRoomsOnly();

    tierManager.forceTier('T3');
    await delay(500);

    const tierEvents = events.filter(e => e.type === 'tier_changed');
    expect(tierEvents.length).toBeGreaterThan(0);
    const first = tierEvents[0];
    if (first.type === 'tier_changed') {
      expect(first.from).toBe('T1');
      expect(first.to).toBe('T3');
    }

    node.stop();
  });
});

// Helpers

function delay(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

function setEquals<T>(a: Set<T>, b: Set<T>): boolean {
  if (a.size !== b.size) return false;
  for (const item of a) {
    if (!b.has(item)) return false;
  }
  return true;
}
